rafx-api/src/backends/gles2/root_signature.rs

```diff
@@ -0,0 +1,445 @@
+    }
+
+    pub(crate) fn uniform_reflection_data(&self) -> &UniformReflectionData {
+        &self.inner.uniform_reflection
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn uniform_index(
+        &self,
+        descriptor_index: RafxDescriptorIndex,
+    ) -> Option<UniformIndex> {
+        self.inner.descriptors[descriptor_index.0 as usize].uniform_index
+    }
+
+    pub fn new(
+        device_context: &RafxDeviceContextGles2,
+        root_signature_def: &RafxRootSignatureDef,
+    ) -> RafxResult<Self> {
+        log::trace!("Create RafxRootSignatureGl");
+
+        //TODO: Count textures and verify we don't exceed max supported
+
+        // If we update this constant, update the arrays in this function
+        assert_eq!(MAX_DESCRIPTOR_SET_LAYOUTS, 4);
+
+        let mut immutable_samplers = vec![];
+        // for sampler_list in root_signature_def.immutable_samplers {
+        //     for sampler in sampler_list.samplers {
+        //         immutable_samplers.push(sampler.clone());
+        //     }
+        // }
+
+        let gl_context = device_context.gl_context();
+
+        // Make sure all shaders are compatible/build lookup of shared data from them
+        let (pipeline_type, merged_resources, _merged_resources_name_index_map) =
+            crate::internal_shared::merge_resources(root_signature_def)?;
+
+        let mut layouts = [
+            DescriptorSetLayoutInfo::default(),
+            DescriptorSetLayoutInfo::default(),
+            DescriptorSetLayoutInfo::default(),
+            DescriptorSetLayoutInfo::default(),
+        ];
+
+        // These are used to populate descriptor_data_offset_in_set in a descriptor, which is later used
+        // to index into Vecs in DescriptorSetArrayData
+        let mut next_descriptor_data_buffer_offset = [0, 0, 0, 0];
+        let mut next_descriptor_data_texture_offset = [0, 0, 0, 0];
+        let mut next_descriptor_data_sampler_offset = [0, 0, 0, 0];
+
+        let mut descriptors = Vec::with_capacity(merged_resources.len());
+        let mut name_to_descriptor_index = FnvHashMap::default();
+
+        let program_ids: Vec<ProgramId> = root_signature_def
+            .shaders
+            .iter()
+            .map(|x| x.gles2_shader().unwrap().gl_program_id())
+            .collect();
+
+        // Lookup for uniform fields
+        let uniform_reflection =
+            UniformReflectionData::new(gl_context, &program_ids, root_signature_def.shaders)?;
+
+        let mut texture_count = 0;
+
+        let mut texture_descriptor_index_sampler_names = Vec::default();
+        let mut sampler_by_gl_name = FnvHashMap::default();
+
+        for resource in &merged_resources {
+            resource.validate()?;
+
+            let descriptor_data_offset_in_set;
+            if resource
+                .resource_type
+                .intersects(RafxResourceType::TEXTURE | RafxResourceType::TEXTURE_READ_WRITE)
+            {
+                //TODO: Handle cube maps
```

Comment:
Cube maps aren't supported with this branch?

Reply:
Not right now. I think it can be added later, this change is already huge. (there's lots of stuff that this branch will never support because of GL ES 2.0)